sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2Writes.scala

```diff
@@ -0,0 +1,187 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import java.util.UUID
+
+import org.apache.spark.SparkException
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst
+import org.apache.spark.sql.catalyst.analysis.Resolver
+import org.apache.spark.sql.catalyst.expressions.{NamedExpression, PredicateHelper, SortOrder}
+import org.apache.spark.sql.catalyst.plans.logical.{AppendData, LogicalPlan, OverwriteByExpression, OverwritePartitionsDynamic, RepartitionByExpression, Sort}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.connector.catalog.Table
+import org.apache.spark.sql.connector.distributions.{ClusteredDistribution, OrderedDistribution, UnspecifiedDistribution}
+import org.apache.spark.sql.connector.expressions.{Expression, FieldReference, IdentityTransform, NullOrdering, SortDirection, SortValue}
+import org.apache.spark.sql.connector.write.{LogicalWriteInfoImpl, RequiresDistributionAndOrdering, SupportsDynamicOverwrite, SupportsOverwrite, SupportsTruncate, Write, WriteBuilder}
+import org.apache.spark.sql.execution.datasources.DataSourceStrategy
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.sources.{AlwaysTrue, Filter}
+
+/**
+ * A rule that constructs [[Write]]s.
+ *
+ * This rule does resolution in the optimizer because some nodes like [[OverwriteByExpression]]
+ * must undergo the expression optimization before we can construct a logical write.
+ */
+object V2Writes extends Rule[LogicalPlan] with PredicateHelper {
+
+  import DataSourceV2Implicits._
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transformDown {
+    case append @ AppendData(relation: DataSourceV2Relation, query, options, _, None) =>
+      val writeBuilder = newWriteBuilder(relation.table, query, options)
+      val write = writeBuilder.build()
+      append.copy(write = Some(write), query = addDistributionAndOrdering(write, query))
+
+    case overwrite @ OverwriteByExpression(
+        relation: DataSourceV2Relation, deleteExpr, query, options, _, None) =>
+      // fail if any filter cannot be converted. correctness depends on removing all matching data.
+      val filters = splitConjunctivePredicates(deleteExpr).flatMap { p =>
+        val filter = DataSourceStrategy.translateFilter(p, supportNestedPredicatePushdown = true)
+        if (filter.isEmpty) {
+          throw new AnalysisException(s"Cannot translate expression to source filter: $p")
+        }
+        filter
+      }.toArray
+
+      val table = relation.table
+      val writeBuilder = newWriteBuilder(table, query, options)
+      val write = writeBuilder match {
+        case builder: SupportsTruncate if isTruncate(filters) =>
+          builder.truncate().build()
+        case builder: SupportsOverwrite =>
```

Comment:
Should we put `SupportsOverwrite` before `SupportsTruncate`? if a builder class extends `SupportsOverwrite` (and `isTruncate` returns true) then it will be matched by the first clause and call the `truncate` method, but we want it to call `overwrite(filters)` right?

Reply:
This is existing logic, just moved.

The reason for this is that `SupportsOverwrite` extends `SupportsTruncate` and calls `overwrite(true)`. Calling `truncate` ensures that the source can implement either one if it chooses. Sometimes truncate may be preferred, and it is easier for a source to receive that call directly rather than writing its own equivalent of `isTruncate`.

Reply:
Got it. I'm just not sure if some data source would extend `SupportsOverwrite` and decides to override/implement different behavior for `overwrite` and `truncate` (which no longer calls `overwrite(true)`, and here we'd pick `truncate` but what they really want is `overwrite`.

Reply:
If the source doesn't implement `truncate`, then it gets `overwrite(true)`: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/SupportsOverwrite.java#L46-L48

Reply:
This logic did not change and should match the previous behavior.